### PR TITLE
add a builtin that can differentiate protocol requirements

### PR DESCRIPTION
**Pull request by marcrasi**
This PR adds a builtin that can differentiate protocol requirements, nearly completing the ability for user code to differentiate protocol requirements! (The only remaining thing is stdlib functions).

I needed to do a few things:
* Added support for differentiating curried functions. (This is necessary because when you reference a protocol requirement from the AST, you get a curried function. The only way to uncurry it is to wrap it in a wrapper, and the wrapper is generic, so we can't differentiate it). This basically meant generalizing the indices structures to work on curried functions and generalizing the `getAutoDiffAssociatedFunctionType`s to work on curried functions. I treat curried functions as fully uncurried when differentiating them.
* Emit VJPs of curry thunks.
* Add some new "Method" builtins that extract associated functions from autodiff functions of type `(T) -> (OtherArgs) -> R`
* Teach SILGenBuiltin to apply curried VJPs.

### CODE REVIEW

File: test/AutoDiff/core_builtins.swift [hunk 1]
```diff
@@ -16,13 +16,44 @@ func evaldiff<T: Differentiable, U: Differentiable>(_ f: @autodiff (T) -> U, _ x
...
+func methoddiff<T: Differentiable, U: Differentiable, R: Differentiable>(_ f: @autodiff (T) -> (U) -> R, _ x: T, _ y: U)
+    -> (R, (T.TangentVector, U.TangentVector) -> R.TangentVector)
+  where T == T.TangentVector, U == U.TangentVector {
+  return Builtin.autodiffApplyMethodJVP(f, x, y)
+}
+
+// CHECK-SIL-LABEL: @{{.*}}methoddiff{{.*}}
+// CHECK-SIL: bb0([[ORIG_RES_BUF:%.*]] : @trivial $*R, [[ORIG_FN:%.*]] : @trivial $@autodiff @noescape @callee_guaranteed (@in_guaranteed T) -> @owned @callee_guaranteed (@in_guaranteed U) -> @out R, [[ORIG_FN_ARG1:%.*]] : @trivial $*T, [[ORIG_FN_ARG2:%.*]] : @trivial $*U):
+// CHECK-SIL:   [[ORIG_FN_ARG1_COPY:%.*]] = alloc_stack $T
+// CHECK-SIL:   copy_addr [[ORIG_FN_ARG1]] to [initialization] [[ORIG_FN_ARG1_COPY]] : $*T
```
Comment (marcrasi) [inline test/AutoDiff/core_builtins.swift]: By the way, while writing this I realized why it emits this copy. It's `SGF.emitRValueAsSingleValue(origFnArgExpr)`.

`emitRValueAsSingleValue` has no idea what we are going to do with the `SILValue` that it produces, so it protects the indirect `@in` argument from being modified by copying it.

There's probably some way to get the original address by promising that we won't write to it.

I'm not going to look for that now though.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Review by rxwei — APPROVED** [review]

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Pull request merged**